# Sobre o notebook

A "hora da prática" é uma atividade do curso de Python para Data Science da Alura que contém diversos desafios envolvendo os aprendizados de cada aula.

O notebook "hora da pratica" é um notebook construído com espaço para as células dos desafios propostos no curso, com ele você tem espaço para construir e executar suas soluções

## Aula 1 - Introdução ao Python

### Questão 1

Imprima a frase `Escola de Dados da Alura!`

In [1]:
print('Escola de dados da alura')

Escola de dados da alura


### Questão 2
Imprima na tela seu nome e seu sobrenome seguindo a estrutura abaixo:
```
Nome: [seu nome]
Sobrenome: [seu sobrenome]
```


In [5]:
nome = 'Jefferson'
sobrenome = 'franca Ferreira'

print(f'Nome: {nome}')
print(f'Sobrenome: {sobrenome}')

Nome: Jefferson
Sobrenome: franca Ferreira


### Questão 3
Imprima o seu primeiro nome letra a letra. Por exemplo, nome é Mirla, então eu obtenho a seguinte saída:
```
M
I
R
L
A
```


In [6]:
print('M')
print('I')
print('R')
print('L')
print('A')

M
I
R
L
A


### Questão 4
Imprima o dia do seu nascimento em formato `dia mês ano`. Os valores de dia e ano não podem estar entre aspas. O formato deve estar como no exemplo, supondo uma data de aniversário dia 28 de fevereiro de 2003:
```
28 fevereiro 2003
```

In [7]:
dia = 28 
mes = 'fevereiro'
ano = 2003

print(f'{dia} {mes} {ano}')

28 fevereiro 2003


### Questão 5
Imprima em um único print o atual ano que você está fazendo esse curso, o valor do ano deve ser um dado numérico. A saída do print deve ser a seguinte:
```
Ano atual: [ano]
```


In [15]:
ano_atual = 2025

print(f'Ano atual: {ano_atual}')

Ano atual: 2025


## Aula 2 - Manipulando dados no Python

### Coleta e amostragem de dados

#### Questão 1
Crie um programa que solicite à pessoa usuária digitar seu nome, e imprima “Olá, [nome]!”.

In [16]:
nome = input ('qual é seu nome; ')

print(f'olá, {nome}!')

qual é seu nome; Jefferson
olá, Jefferson!


#### Questão 2
Crie um programa que solicite à pessoa usuária digitar seu nome e idade, e imprima “Olá, [nome], você tem [idade] anos.”.

In [17]:
nome = input ('Qual é sue nome: ')
idade = 37

print(f'Olá, {nome}, você tem {idade}, anos.')

Qual é sue nome: jefferson
Olá, jefferson, você tem 37, anos.


#### Questão 3
Crie um programa que solicite à pessoa usuária digitar seu nome, idade e altura em metros, e imprima “Olá, [nome], você tem [idade] anos e mede [altura] metros!”.

In [19]:
nome = input ('Qual é seu nome')
idade = 37
mede = 1.75

print(f'Olá, {nome}, você tem {idade}, e mede {mede}, metros!')

Qual é seu nomejefferson
Olá, jefferson, você tem 37, e mede 1.75, metros!


### Calculadora com operadores

#### Questão 1

Crie um programa que solicite dois valores numéricos à pessoa usuária e imprima a soma dos dois valores.

In [21]:
p1 = 10
p2 = 56
print (p1 + p2)

66


#### Questão 2
Crie um programa que solicite três valores numéricos à pessoa usuária e imprima a soma dos três valores.

In [23]:
p1 = 24
p2 = 35
p3 = 56
print (p1 + p2 + p3)

115


#### Questão 3
Crie um programa que solicite dois valores numéricos à pessoa usuária e imprima a subtração do primeiro pelo o segundo valor.

In [24]:
P1 = 24
p2 = 12
print(p1 - p2)

12


#### Questão 4

Crie um programa que solicite dois valores numéricos à pessoa usuária e imprima a multiplicação dos dois valores.


In [25]:
P1 = 87
p2 = 100
print(p1 + p2)

124


#### Questão 5

Crie um programa que solicite dois valores numéricos, um numerador e um denominador, e realize a divisão entre os dois valores. Deixe claro que o valor do denominador não pode ser `0`.


In [28]:
numerador = float(input('Digite o numerador: '))
denominador = float(input('Digite denominador(não pede ser 0): '))

if denominador !=0:
    resultado = numerador / denominador
    print(f'O denominador da divisão é: {resultado}')
else:
    print('Erro: O denominador não pode ser 0.')

Digite o numerador: 10
Digite denominador(não pede ser 0): 0
Erro: O denominador não pode ser 0.


#### Questão 6

Crie um programa que solicite dois valores numéricos, um operador e uma potência, e realize a exponenciação entre esses dois valores.


In [31]:
base = float(input('Digite o número base: '))
expoente = float(input('Digite o número expoente: '))

resultado = (base ** expoente)

print(f'O resultado de {base} elevado a {expoente} é: {resultado}')


Digite o número base: 10
Digite o número expoente: 10
O resultado de 10.0 elevado a 10.0 é: 10000000000.0


#### Questão 7
Crie um programa que solicite dois valores numéricos, um numerador e um denominador e realize a divisão inteira entre os dois valores. Deixe claro que o valor do denominador não pode ser `0`.

In [38]:
numerador = int(input('Digite o número numerador: '))
denominador = int(input('Digite o número denominador: '))

if denominador !=0:
    resultado = int(numerador / denominador)
    print(f'o resultado do numerador é: {resultado}')
else:
    print(f'Erro: O denominador não pode ser 0.')

Digite o número numerador: 24
Digite o número denominador: 5
o resultado do numerador é: 4


#### Questão 8

Crie um programa que solicite dois valores numéricos, um numerador e um denominador, e retorne o resto da divisão entre os dois valores. Deixe claro que o valor do denominador não pode ser `0`.

In [3]:
numerador = float(input('Digite o número numerador: '))
denominador = float(input('Digite o número denominador: '))

if denominador != 0:
    resto = numerador % denominador
    print(f'O resto da divisão entre {numerador} e {denominador} é {resto}.')
else:
    print('Erro: o denominador não poder ser zero.')

Digite o número numerador: 17
Digite o número denominador: 1
O resto da divisão entre 17.0 e 1.0 é 0.0.


#### Questão 9

Crie um código que solicita 3 notas de um estudante e imprima a média das notas.

In [6]:
nota_1 = float(input('Digite 1° nota: '))
nota_2 = float(input('Digite 2° nota: '))
nota_3 = float(input('Digite 3° nota: '))

media = (nota_1 + nota_2 + nota_3) / 3

print(f'Média da nota do estudante é: {media: .2f}')

Digite 1° nota: 7
Digite 2° nota: 5
Digite 3° nota: 9
Média da nota do estudante é:  7.00


#### Questão 10

Crie um código que calcule e imprima a média ponderada dos números 5, 12, 20 e 15 com pesos respectivamente iguais a 1, 2, 3 e 4.

In [11]:
numeros = [5, 12, 20, 15]
pesos = [1, 2, 3, 4]

somatorio_ponderado = sum(numero * peso for numero, peso in zip(numeros, pesos))

soma_pesos = sum(pesos)
media_ponderada = somatorio_ponderado / soma_pesos

print(f'A média ponderada é: {media_ponderada: .2f}')

A média ponderada é:  14.90


### Editando textos

#### Questão 1

Crie uma variável chamada “`frase`” e atribua a ela uma string de sua escolha. Em seguida, imprima a frase na tela.

In [12]:
frase = 'A tecnologia aproxima pessoas e possibilidades.'
frase

'A tecnologia aproxima pessoas e possibilidades.'

#### Questão 2
Crie um código que solicite uma frase e depois imprima a frase na tela.

In [19]:
frase = input('Digete uma frase: ')
print(f'A frase digitada foi: {frase}')

Digete uma frase: A tecnologia aproxima pessoas e possibilidades.
A frase digitada foi: A tecnologia aproxima pessoas e possibilidades.


#### Questão 3

Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase digitada mas com todas as letras maiúsculas.

In [20]:
frase = 'A tecnologia aproxima pessoas e possibilidades.'
print(f'Frase toda maiúsculas: {frase.upper()}')

Frase toda maiúsculas: A TECNOLOGIA APROXIMA PESSOAS E POSSIBILIDADES.


#### Questão 4

Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase digitada mas com todas as letras minúsculas.

In [22]:
frase = 'A tecnologia aproxima pessoas e possibilidades.'
print(f'Frase toda minúsculas: {frase.lower()}')

Frase toda minúsculas: a tecnologia aproxima pessoas e possibilidades.


#### Questão 5

Crie uma variável chamada “`frase`” e atribua a ela uma string de sua escolha. Em seguida, imprima a frase sem espaços em branco no início e no fim.

In [24]:
frase.strip()

'A tecnologia aproxima pessoas e possibilidades.'

#### Questão 6

Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase sem espaços em branco no início e no fim.

In [25]:
frase = input('Digite uma frase: ')

frase_sem_espacos = frase.strip()

print(f'A frase sem espaço em branco é: {frase_sem_espacos}')

Digite uma frase: Olá, mundo!
A frase sem espaço em branco é: Olá, mundo!


#### Questão 7


Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase sem espaços em branco no início e no fim e em letras minúsculas.

In [32]:
frase = input('Digite uma frase: ')

frase_formatada = frase.strip().lower()


print(f'A frase formatada é: {frase_formatada}')

Digite uma frase: Olá, Mundo!
A frase formatada é: olá, mundo!


#### Questão 8

Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase com todas as vogais “`e`” trocadas pela letra “`f`”.

In [34]:
frase = 'A tecnologia aproxima pessoas e possibilidades'
print(frase.replace('e', 'f'))

A tfcnologia aproxima pfssoas f possibilidadfs


#### Questão 9


Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase com todas as vogais “`a`” trocadas pela caractere  “`@`”.

In [35]:
print(frase.replace('a', '@'))

A tecnologi@ @proxim@ pesso@s e possibilid@des


#### Questão 10


Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase com todas as consoantes “`s`” trocadas pelo caractere  “`$`”.

In [36]:
print(frase.replace('s', '$'))

A tecnologia aproxima pe$$oa$ e po$$ibilidade$


## Aula 3 - Estruturas condicionais

### Aquecendo na programação

#### Questão 1
Escreva um programa que peça à pessoa usuária para fornecer dois números e exibir o número maior.

In [4]:
numero_1 = float(input('Digite o primeiro número: '))
numero_2 = float(input('Digite o segundo número: '))

if numero_1 > numero_2:
    print(f'O maior número é {numero_1}')
elif numero_2 > numero_1:
    print(f'O maior número é {numero_2}')
else:
    print('Os números são iguais.')

Digite o primeiro número: 6
Digite o segundo número: 6
Os números são iguais.


#### Questão 2
Escreva um programa que solicite o percentual de crescimento de produção de uma empresa e informe se houve um crescimento (porcentagem positiva) ou decrescimento (porcentagem negativa).

In [7]:
percentual_crescimento = float(input('Digite o percentual de crescimento da produçao (pode ser negativo): '))

if percentual_crescimento > 0:
    print('Houve um crescimento na produção.')
elif percentual_crescimento < 0:
    print('Houve um decrescimento na produção.')
else:
    print('Não huve crescimento nem decrescimento na produçõ.')

Digite o percentual de crescimento da produçao (pode ser negativo): 0
Não huve crescimento nem decrescimento na produçõ.


#### Questão 3
Escreva um programa que determine se uma letra fornecida pela pessoa usuária é uma vogal ou consoante.

In [9]:
# Solicitar ao usuário uma letra
letra = input("Digite uma letra: ").lower()

# Verificar se a entrada é válida (apenas uma letra)
if len(letra) == 1 and letra.isalpha():
    
    # Determinar se é uma vogal ou consoante
    if letra in 'aeiou':
        print("A letra fornecida é uma vogal.")
    else:
        print("A letra fornecida é uma consoante.")
else:
    print("Por favor, digite apenas uma letra válida.")


Digite uma letra: j
A letra fornecida é uma consoante.


#### Questão 4
Escreva um programa que leia valores médios de preços de um modelo de carro por 3 anos consecutivos e exiba o valor mais alto e mais baixo entre esses três anos.

In [15]:
# Solicitar ao usuário os valores médios dos preços para cada ano
preco_ano1 = float(input("Digite o valor médio do preço no 1º ano: "))
preco_ano2 = float(input("Digite o valor médio do preço no 2º ano: "))
preco_ano3 = float(input("Digite o valor médio do preço no 3º ano:"))

# Encontrar o valor mais alto e mais baixo
valor_maximo = max(preco_ano1, preco_ano2, preco_ano3)
valor_minimo = min(preco_ano1, preco_ano2, preco_ano3)

# Exibir os resultados
print(f"O valor mais alto entre os três anos é: {valor_maximo}")
print(f"O valor mais baixo entre os três anos é: {valor_minimo}")



Digite o valor médio do preço no 1º ano: 0
Digite o valor médio do preço no 2º ano: 0
Digite o valor médio do preço no 3º ano:0
O valor mais alto entre os três anos é: 0.0
O valor mais baixo entre os três anos é: 0.0


#### Questão 5
Escreva um programa que pergunte sobre o preço de três produtos e indique qual é o produto mais barato para comprar.

In [16]:
# Solicitar ao usuário os preços dos três produtos
preco1 = float(input("Digite o preço do primeiro produto: R$ "))
preco2 = float(input("Digite o preço do segundo produto: R$ "))
preco3 = float(input("Digite o preço do terceiro produto: R$ "))

# Determinar qual produto é o mais barato
if preco1 < preco2 and preco1 < preco3:
    print("Você deve comprar o primeiro produto, que é o mais barato.")
elif preco2 < preco1 and preco2 < preco3:
    print("Você deve comprar o segundo produto, que é o mais barato.")
elif preco3 < preco1 and preco3 < preco2:
    print("Você deve comprar o terceiro produto, que é o mais barato.")
else:
    print("Há mais de um produto com o menor preço.")


Digite o preço do primeiro produto: R$ 30
Digite o preço do segundo produto: R$ 190
Digite o preço do terceiro produto: R$ 220
Você deve comprar o primeiro produto, que é o mais barato.


#### Questão 6
Escreva um programa que leia três números e os exiba em ordem decrescente.

In [17]:
# Solicitar ao usuário que insira três números
numero1 = float(input("Digite o primeiro número: "))
numero2 = float(input("Digite o segundo número: "))
numero3 = float(input("Digite o terceiro número: "))

# Criar uma lista com os números e ordená-los em ordem decrescente
numeros = [numero1, numero2, numero3]
numeros.sort(reverse=True)

# Exibir os números em ordem decrescente
print("Os números em ordem decrescente são:")
for numero in numeros:
    print(numero)

Digite o primeiro número: 10
Digite o segundo número: 90
Digite o terceiro número: 120
Os números em ordem decrescente são:
120.0
90.0
10.0


#### Questão 7
Escreva um programa que pergunte em qual turno a pessoa usuária estuda ("manhã", "tarde" ou "noite") e exiba a mensagem "Bom Dia!", "Boa Tarde!", "Boa Noite!", ou "Valor Inválido!", conforme o caso.

In [20]:
# Solicitar ao usuário em qual turno estuda
turno = input("Em qual turno você estuda? (manhã/tarde/noite): ").lower()

# Exibir a mensagem apropriada
if turno == "manhã":
    print("Bom Dia!")
elif turno == "tarde":
    print("Boa Tarde!")
elif turno == "noite":
    print("Boa Noite!")
else:
    print("Valor Inválido!")

Em qual turno você estuda? (manhã/tarde/noite): manhã
Bom Dia!


#### Questão 8
Escreva um programa que peça um número inteiro à pessoa usuária e determine se ele é par ou ímpar. Dica: Você pode utilizar o operador módulo `%`.

In [22]:
# Solicitar ao usuário um número inteiro
numero = int(input("Digite um número inteiro: "))

# Determinar se o número é par ou ímpar
if numero % 2 == 0:
    print("O número é par.")
else:
    print("O número é ímpar.")

Digite um número inteiro: 3
O número é ímpar.


#### Questão 9
Escreva um programa que peça um número à pessoa usuária e informe se ele é inteiro ou decimal.

In [27]:
# Solicitar ao usuário que insira um número
numero = input("Digite um número: ")

# Verificar se o número é inteiro ou decimal
try:
    if "." in numero:  # Verifica se há um ponto decimal na entrada
        float(numero)  # Tenta converter para float
        print("O número é decimal.")
    else:
        int(numero)  # Tenta converter para inteiro
        print("O número é inteiro.")
except ValueError:
    print("A entrada não é um número válido.")

Digite um número: 3
O número é inteiro.


### Momento dos projetos

#### Questão 10
Um programa deve ser escrito para ler dois números e, em seguida, perguntar à pessoa usuária qual operação ele deseja realizar. O resultado da operação deve incluir informações sobre o número - se é par ou ímpar, positivo ou negativo e inteiro ou decimal.

In [28]:
# Funções auxiliares para verificar informações sobre os números
def verificar_par_ou_impar(numero):
    return "par" if numero % 2 == 0 else "ímpar"

def verificar_positivo_ou_negativo(numero):
    return "positivo" if numero > 0 else "negativo" if numero < 0 else "neutro"

def verificar_inteiro_ou_decimal(numero):
    return "inteiro" if numero == int(numero) else "decimal"

# Solicitar ao usuário dois números
numero1 = float(input("Digite o primeiro número: "))
numero2 = float(input("Digite o segundo número: "))

# Perguntar ao usuário qual operação ele deseja realizar
print("Escolha uma operação: ")
print("1 - Soma")
print("2 - Subtração")
print("3 - Multiplicação")
print("4 - Divisão")
operacao = int(input("Digite o número correspondente à operação desejada: "))

# Realizar a operação escolhida
if operacao == 1:
    resultado = numero1 + numero2
    operacao_str = "Soma"
elif operacao == 2:
    resultado = numero1 - numero2
    operacao_str = "Subtração"
elif operacao == 3:
    resultado = numero1 * numero2
    operacao_str = "Multiplicação"
elif operacao == 4:
    if numero2 != 0:
        resultado = numero1 / numero2
        operacao_str = "Divisão"
    else:
        print("Erro: divisão por zero não é permitida.")
        exit()
else:
    print("Operação inválida.")
    exit()

# Exibir o resultado e informações sobre o número
print(f"\nResultado da operação ({operacao_str}): {resultado}")
print(f"O resultado é um número {verificar_par_ou_impar(resultado)}, {verificar_positivo_ou_negativo(resultado)} e {verificar_inteiro_ou_decimal(resultado)}.")

Digite o primeiro número: 40
Digite o segundo número: 58
Escolha uma operação: 
1 - Soma
2 - Subtração
3 - Multiplicação
4 - Divisão
Digite o número correspondente à operação desejada: 4

Resultado da operação (Divisão): 0.6896551724137931
O resultado é um número ímpar, positivo e decimal.


#### Questão 11
Escreva um programa que peça à pessoa usuária três números que representam os lados de um triângulo. O programa deve informar se os valores podem ser utilizados para formar um triângulo e, caso afirmativo, se ele é equilátero, isósceles ou escaleno. Tenha em mente algumas dicas:
- Três lados formam um triângulo quando a soma de quaisquer dois lados for maior que o terceiro;
- Triângulo Equilátero: três lados iguais;
- Triângulo Isósceles: quaisquer dois lados iguais;
- Triângulo Escaleno: três lados diferentes;


In [29]:
# Solicitar ao usuário os três lados do triângulo
lado1 = float(input("Digite o valor do primeiro lado: "))
lado2 = float(input("Digite o valor do segundo lado: "))
lado3 = float(input("Digite o valor do terceiro lado: "))

# Verificar se os lados podem formar um triângulo
if (lado1 + lado2 > lado3) and (lado1 + lado3 > lado2) and (lado2 + lado3 > lado1):
    print("Os valores podem formar um triângulo.")
    
    # Verificar o tipo de triângulo
    if lado1 == lado2 == lado3:
        print("O triângulo é equilátero.")
    elif lado1 == lado2 or lado1 == lado3 or lado2 == lado3:
        print("O triângulo é isósceles.")
    else:
        print("O triângulo é escaleno.")
else:
    print("Os valores não podem formar um triângulo.")

Digite o valor do primeiro lado: 20
Digite o valor do segundo lado: 29
Digite o valor do terceiro lado: 10
Os valores podem formar um triângulo.
O triângulo é escaleno.


#### Questão 12
Um estabelecimento está vendendo combustíveis com descontos variados. Para o etanol, se a quantidade comprada for até 15 litros, o desconto será de 2% por litro. Caso contrário, será de 4% por litro. Para o diesel, se a quantidade comprada for até 15 litros, o desconto será de 3% por litro. Caso contrário, será de 5% por litro. O preço do litro de diesel é R\$ 2,00 e o preço do litro de etanol é R\$ 1,70. Escreva um programa que leia a quantidade de litros vendidos e o tipo de combustível (E para etanol e D para diesel) e calcule o valor a ser pago pelo cliente. Tenha em mente algumas dicas:

- O do valor do desconto será a multiplicação entre preço do litro, quantidade de litros e o valor do desconto.
- O valor a ser pago por um cliente será o resultado da multiplicação do preço do litro pela quantidade de litros menos o valor de desconto resultante do cálculo.


In [30]:
# Preços dos combustíveis por litro
preco_etanol = 1.70
preco_diesel = 2.00

# Solicitar ao usuário a quantidade de litros e o tipo de combustível
quantidade_litros = float(input("Digite a quantidade de litros vendidos: "))
tipo_combustivel = input("Digite o tipo de combustível (E para etanol, D para diesel): ").upper()

# Inicializar variáveis de preço e desconto
preco_litro = 0
desconto = 0

# Calcular o desconto e o valor a pagar com base no tipo de combustível e quantidade
if tipo_combustivel == "E":
    preco_litro = preco_etanol
    if quantidade_litros <= 15:
        desconto = 0.02  # 2% de desconto por litro
    else:
        desconto = 0.04  # 4% de desconto por litro
elif tipo_combustivel == "D":
    preco_litro = preco_diesel
    if quantidade_litros <= 15:
        desconto = 0.03  # 3% de desconto por litro
    else:
        desconto = 0.05  # 5% de desconto por litro
else:
    print("Tipo de combustível inválido.")
    exit()

# Calcular o valor do desconto e o valor final a pagar
valor_desconto = preco_litro * quantidade_litros * desconto
valor_a_pagar = (preco_litro * quantidade_litros) - valor_desconto

# Exibir o resultado
print(f"Valor do desconto: R$ {valor_desconto:.2f}")
print(f"Valor a pagar: R$ {valor_a_pagar:.2f}")

Digite a quantidade de litros vendidos: 10.000
Digite o tipo de combustível (E para etanol, D para diesel): E
Valor do desconto: R$ 0.34
Valor a pagar: R$ 16.66


#### Questão 13
Em uma empresa de venda de imóveis você precisa criar um código que analise os dados de vendas anuais para ajudar a diretoria na tomada de decisão. O código precisa coletar os dados de quantidade de venda durante os anos de 2022 e 2023 e fazer um cálculo de variação percentual. A partir do valor da variação, deve ser enviada às seguintes sugestões:
- Para variação acima de 20%: bonificação para o time de vendas.
- Para variação entre 2% e 20%: pequena bonificação para time de vendas.
- Para variação entre 2% e -10%: planejamento de políticas de incentivo às vendas.
- Para bonificações abaixo de -10%: corte de gastos.


In [31]:
# Solicitar as quantidades de vendas dos anos de 2022 e 2023
vendas_2022 = float(input("Digite a quantidade de vendas no ano de 2022: "))
vendas_2023 = float(input("Digite a quantidade de vendas no ano de 2023: "))

# Calcular a variação percentual
variacao_percentual = ((vendas_2023 - vendas_2022) / vendas_2022) * 100

# Exibir a variação percentual
print(f"Variação percentual: {variacao_percentual:.2f}%")

# Enviar sugestões com base na variação
if variacao_percentual > 20:
    print("Sugestão: Bonificação para o time de vendas.")
elif 2 <= variacao_percentual <= 20:
    print("Sugestão: Pequena bonificação para o time de vendas.")
elif -10 <= variacao_percentual < 2:
    print("Sugestão: Planejamento de políticas de incentivo às vendas.")
else:
    print("Sugestão: Corte de gastos.")

Digite a quantidade de vendas no ano de 2022: 23.000
Digite a quantidade de vendas no ano de 2023: 57.000
Variação percentual: 147.83%
Sugestão: Bonificação para o time de vendas.


## Aula 4 - Estruturas de repetição

### Aquecendo na programação

#### Questão 1
Escreva um programa que peça dois números inteiros e imprima todos os números inteiros entre eles.

#### Questão 2
Escreva um programa para calcular quantos dias levará para a colônia de uma bactéria A ultrapassar ou igualar a colônia de uma bactéria B, com base nas taxas de crescimento de 3% e 1,5% respectivamente. Considere que a colônia A inicia com 4 elementos e a colônia B com 10 elementos.

#### Questão 3
Para tratar uma quantidade de 15 dados de avaliações de pessoas usuárias de um serviço da empresa, precisamos verificar se as notas são válidas. Então, escreva um programa que vai receber a nota de 0 a 5 de todos os dados e verificar se é um valor válido. Caso seja inserido uma nota acima de 5 ou abaixo de 0, repita até que a pessoa usuária insira um valor válido.

#### Questão 4
Desenvolva um programa que leia um conjunto indeterminado de temperaturas em Celsius e faça uma análise. Portanto, escreva um programa que leia temperaturas e informe a média delas. A leitura deve ser encerrada ao ser enviado o valor -273°C.

#### Questão 5
Escreva um programa que calcule o fatorial de um número inteiro fornecido pela pessoa usuária. O fatorial de um número inteiro é a multiplicação desse número por todos os seus antecessores até o número 1. Por exemplo, o fatorial de 5 é 5 x 4 x 3 x 2 x 1 = 120.

### Momento dos projetos

#### Questão 6
Escreva um programa que gere a tabuada de um número inteiro de 1 a 10, de acordo com a escolha da pessoa usuária. Como exemplo, para o número 2, a tabuada deve ser mostrada no seguinte formato:
```
Tabuada do 2:
2 x 1 = 2
2 x 2 = 4
[...]
2 x 10 = 20
```

#### Questão 7
Os números primos possuem várias aplicações dentro da Ciência de Dados, por exemplo, na criptografia e segurança. Um número primo é aquele que é divisível apenas por um e por ele mesmo. Faça um programa que peça um número inteiro e determine se ele é ou não um número primo.

#### Questão 8
Vamos entender a distribuição de idades de pensionistas de uma empresa de previdência. Escreva um programa que leia as idades de uma quantidade não informada de clientes e mostre a distribuição em intervalos de [0-25], [26-50], [51-75] e [76-100]. Encerre a entrada de dados com um número negativo.

#### Questão 9
Em uma eleição para gerência em uma empresa com 20 funcionários, existem quatro candidatos. Escreva um programa que calcule o vencedor da eleição. A votação ocorreu da seguinte maneira:
- Cada funcionário votou em um dos quatro candidatos (representados pelos números 1, 2, 3 e 4).
- Também foram contabilizados os votos nulos (representado pelo número 5) e os votos em branco (representado pelo número 6).

Ao final da votação, o programa deve exibir o total de votos para cada candidato, o número de votos nulos e o número de votos em branco. Além disso, deve calcular e exibir a porcentagem de votos nulos em relação ao total de votos e a porcentagem de votos em branco em relação ao total de votos.

## Aula 5 - Estruturas de dados

### Aquecendo na programação

#### Questão 1
Faça um programa que tenha a seguinte lista contendo os valores de gastos de uma empresa de papel `[2172.54, 3701.35, 3518.09, 3456.61, 3249.38, 2840.82, 3891.45, 3075.26, 2317.64, 3219.08]`. Com esses valores, faça um programa que calcule a média de gastos. Dica, use as funções built-in [`sum()`](https://docs.python.org/3/library/functions.html#sum) e [`len()`](https://docs.python.org/3/library/functions.html#len).

#### Questão 2
 Com os mesmos dados da questão anterior, defina quantas compras foram acima de 3000 reais e calcule a porcentagem quanto ao total de compras.

#### Questão 3
Faça um código que colete em uma lista 5 números inteiros quaisquer e imprima a lista. Exemplo: `[1,4,7,2,4]`.

#### Questão 4
Colete novamente 5 inteiros e imprima a lista em ordem inversa à enviada.

#### Questão 5
Faça um programa que, ao inserir um número qualquer, criará uma lista contendo todos os números primos entre 1 e o número digitado.

#### Questão 6
Escreva um programa que peça uma data informando o dia, mês e ano e determine se ela é uma data válida para uma análise.

### Momento dos projetos

#### Questão 7
Para um estudo envolvendo o nível de multiplicação de bactérias em uma colônia, foi coletado o número de bactérias multiplicadas por dia, com base em um processo de duplicação diária e pode ser observado a seguir: `[1.2, 2.1, 3.3, 5.0, 7.8, 11.3, 16.6, 25.1, 37.8, 56.9]`. Tendo esses valores, faça um código que gere uma lista contendo o percentual de crescimento de bactérias por dia, comparando o número de bactérias em cada dia com o número de bactérias do dia anterior. Dica: para calcular o percentual de crescimento usamos a seguinte equação: `100 * (amostra_atual - amostra_passada) / (amostra_passada)`.

#### Questão 8
Para uma seleção de produtos alimentícios, precisamos separar o conjunto de IDs dados por números inteiros, sabendo que os produtos com ID par são doces e os com ID ímpar são amargos. Monte um código que colete 10 IDs. Depois, calcule e mostre a quantidade de produtos doces e amargos.

#### Questão 9
Desenvolva um programa que informa a nota de um aluno de acordo com suas respostas. Ele deve pedir a resposta de um aluno para cada questão e é preciso verificar se a resposta foi igual ao gabarito. Cada questão vale um ponto e existem alternativas:  A, B, C ou D.
```
Gabarito da Prova:
01 - D
02 - A
03 - C
04 - B
05 - A
06 - D
07 - C
08 - C
09 - A
10 - B
```

#### Questão 10
Um instituto de meteorologia deseja fazer um estudo de temperatura média de cada mês do ano. Para isso, você precisa fazer um código que colete e armazene essas temperaturas médias em uma lista. Depois, calcule a média anual das temperaturas e mostre todas as temperaturas acima da média anual, e em que mês elas ocorreram, mostrando os meses por extenso: Janeiro, Fevereiro, etc.

#### Questão 11
Uma empresa de e-commerce está interessada em analisar as vendas dos seus produtos. Os dados das vendas foram armazenados em um dicionário:
```
{'Produto A': 300, 'Produto B': 80, 'Produto C': 60,
 'Produto D': 200, 'Produto E': 250, 'Produto F': 30}
```
Escreva um código que calcule o total de vendas e o produto mais vendido.

#### Questão 12
Uma pesquisa de mercado foi feita para decidir qual design de uma marca infantil mais agrada crianças. A pesquisa foi feita e o votos computados podem ser observados abaixo:
```
'''
Tabela de votos da marca
Design 1 - 1334 votos
Design 2 - 982 votos
Design 3 - 1751 votos
Design 4 - 210 votos
Design 5 - 1811 votos
'''
```
Adapte os dados fornecidos a você para uma estrutura de dicionário e a partir dele, informe o design vencedor e a porcentagem de votos recebidos.


#### Questão 13
Os funcionários de um setor da empresa que você trabalha vão receber um abono correspondente a 10% do seu salário devido ao ótimo desempenho do time. O setor financeiro solicitou sua ajuda para a verificação das consequências financeiras que esse abono irá gerar nos recursos. Assim, foi encaminhada para você uma lista com os salários que receberão o abono: `[1172, 1644, 2617, 5130, 5532, 6341, 6650, 7238, 7685, 7782, 7903]`. O abono de cada funcionário não pode ser inferior a 200. Em código, transforme cada um dos salários em chaves de um dicionário e o abono de cada salário no elemento. Depois, informe o total de gastos com o abono, quantos funcionários receberam o abono mínimo e qual o maior valor de abono fornecido.


#### Questão 14
Uma equipe de cientistas de dados está estudando a diversidade biológica em uma floresta. A equipe fez a coleta de informações sobre o número de espécies de plantas e animais em cada área da floresta e armazenaram essas informações em um dicionário. Nele, a chave descreve a área dos dados e os valores nas listas correspondem às espécies de plantas e animais nas áreas, respectivamente.
```
{'Área Norte': [2819, 7236],
 'Área Leste': [1440, 9492],
 'Área Sul': [5969, 7496],
 'Área Oeste': [14446, 49688],
 'Área Centro': [22558, 45148]}
```
Escreva um código para calcular a média de espécies por área e identificar a área com a maior diversidade biológica. Dica: use as funções built-in [`sum()`](https://docs.python.org/3/library/functions.html#sum) e [`len()`](https://docs.python.org/3/library/functions.html#len).


#### Questão 15
O setor de RH da sua empresa te pediu uma ajuda para analisar as idades dos funcionários de 4 setores da empresa. Para isso, ele te forneceu os seguintes dados:
```
{'Setor A': [22, 26, 30, 30, 35, 38, 40, 56, 57, 65],
 'Setor B': [22, 24, 26, 33, 41, 49, 50, 54, 60, 64],
 'Setor C': [23, 26, 26, 29, 34, 35, 36, 41, 52, 56],
 'Setor D': [19, 20, 25, 27, 34, 39, 42, 44, 50, 65]}
``` 
Sabendo que cada setor tem 10 funcionários, construa um código que calcule a média de idade de cada setor, a idade média geral entre todos os setores e quantas pessoas estão acima da idade média geral.
